### PR TITLE
Adding a new check - frequent substring

### PR DESCRIPTION
**Pull request by shiritdvir**
**A new check - frequent substrings:**

This check detects frequent substrings (n-grams, with 4 by default)  in the dataset and alerts if the number of the returned substrings crosses the `min_substrings` parameter (default 1).

### CODE REVIEW

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+# Copyright (C) 2021-2023 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Module contains Frequent Substrings check."""
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:11]:
```suggestion
"""Module containing the Frequent Substrings check."""
```

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+from deepchecks.core.errors import DeepchecksValueError
+from deepchecks.nlp import Context, SingleDatasetCheck
+from deepchecks.nlp._shared_docs import docstrings
+from deepchecks.nlp.text_data import TextData
+from deepchecks.utils.strings import format_percent
+
+__all__ = ['FrequentSubstrings']
+
+
+@docstrings
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:28]: if you are not using the {} doctring insertions, you don't need this decorator

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+class FrequentSubstrings(SingleDatasetCheck):
+    """Checks for frequent substrings in the dataset.
+
+    Parameters
+    ----------
+    n_to_show : int, default: 5
+        number of most common duplicated samples to show.
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:35]: Update the description here to match the new usage of this parameter(:

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    n_samples : int, default: 10_000_000
+        number of samples to use for this check.
+    random_state : int, default: 42
+        random seed for all check internals.
+    min_threshold: float, default: 0.05
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:40]: We need to find a way to make it clearer that this is a threshold on the ratio of samples the substrings appears in. 
Perhaps `min_substring_ratio`?
Need to update the explanation of this argument to explain this fact

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+from deepchecks.nlp._shared_docs import docstrings
+from deepchecks.nlp.text_data import TextData
+from deepchecks.utils.strings import format_percent
+
+__all__ = ['FrequentSubstrings']
+
+
+@docstrings
+class FrequentSubstrings(SingleDatasetCheck):
+    """Checks for frequent substrings in the dataset.
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:30]: Add below this a short description of the algorithm - we search for n-grams that appear more than x, then search for peaks...

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    n_to_show : int, default: 5
+        number of most common duplicated samples to show.
+    n_samples : int, default: 10_000_000
+        number of samples to use for this check.
+    random_state : int, default: 42
+        random seed for all check internals.
+    min_threshold: float, default: 0.05
+        minimum threshold for tagging a substring as frequent
+    min_ngram: int, default: 4
+        minimum n for the n-gram extraction
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:43]: Elaborate a bit more here. E.g. "Minimum amount of words for a substring to be considered a frequent substring"

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+@docstrings
+class FrequentSubstrings(SingleDatasetCheck):
+    """Checks for frequent substrings in the dataset.
+
+    Parameters
+    ----------
+    n_to_show : int, default: 5
+        number of most common duplicated samples to show.
+    n_samples : int, default: 10_000_000
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:36]: would reduce to either  100K / 1m depends how fast or slow this check is

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        number of samples to use for this check.
+    random_state : int, default: 42
+        random seed for all check internals.
+    min_threshold: float, default: 0.05
+        minimum threshold for tagging a substring as frequent
+    min_ngram: int, default: 4
+        minimum n for the n-gram extraction
+    epsilon: float, default: 0.02
+        maximum difference for filtering long substrings
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:45]: Also here, can you explain what is the "difference" here?

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+    significant_threshold: float, default: 0.3
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:46]: Change this to align with the new naming you chose for min_threshold

Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:46]: Would rephrase a bit, from reading the docstring i am not sure what is the difference between significant_threshold and min_threshold. Also epsilon isnt clear enough

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+                'Frequency': [item[1]['freq'] for item in sorted_substrings],
+                'Sample IDs': [item[1]['original_indexes'] for item in sorted_substrings]
+            })
+
+            df = self._filter_significant(df)
+            df['Number of Samples'] = df['Sample IDs'].str.len()
+
+            value = df
+            percent_of_frequent = sum(df['Number of Samples'])
+            display = [
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:199]: add here
if context.with_display:
 display = none
else:
...

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        for item in data:
+            for ngram in self._extract_ngrams(item[1], n):
+                ngrams.add(ngram)
+
+        for ngram in ngrams:
+            for index, item in enumerate(data):
+                if ngram in item[1]:
+                    ngram_info[ngram]['freq'] += 1
+                    ngram_info[ngram]['original_indexes'].add(item[0])
+                    ngram_info[ngram]['new_indexes'].add(index)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:99]: Those two for loop can be combined - only need to do one pass on the data

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    @staticmethod
+    def _get_peak_cut_ind(df, min_diff=0.05):
+        df['diff'] = df['Frequency'].diff()
+        df[df['diff'] <= min_diff]['diff'] = 0
+        max_peak = df['diff'].min()
+        if max_peak == 0:
+            return df
+        return list(df[df['diff'] == max_peak].index)[0]
+
+    def _get_ngram_frequencies_and_indexes(self, data, n, num_samples):
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:85]: add short docstring stating what is the return value of this + few comments in relevant section of the fucntion

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+        ngrams = set()
+        for item in data:
+            for ngram in self._extract_ngrams(item[1], n):
+                ngrams.add(ngram)
+
+        for ngram in ngrams:
+            for index, item in enumerate(data):
+                if ngram in item[1]:
+                    ngram_info[ngram]['freq'] += 1
+                    ngram_info[ngram]['original_indexes'].add(item[0])
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:98]: I think you can join the logic of the previous loop, _extract_ngrams and this loop to do the whole ngram extraction. +freq extraction + original indexes extraction in one loop pass by using ngram_info from the start.

**Review by Nadav-Barak — COMMENTED** [review]
Added some comments - mainly try and make the logic more clear

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+    def _get_peak_cut_ind(df, min_diff=0.05):
+        df['diff'] = df['Frequency'].diff()
+        df[df['diff'] <= min_diff]['diff'] = 0
+        max_peak = df['diff'].min()
+        if max_peak == 0:
+            return df
+        return list(df[df['diff'] == max_peak].index)[0]
+
+    def _get_ngram_frequencies_and_indexes(self, data, n, num_samples):
+        ngram_info = defaultdict(lambda: {'freq': 0, 'original_indexes': set(), 'new_indexes': set()})
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:86]: naming-wise - why are these indexes "new"?

File: tests/nlp/checks/data_integrity/frequent_substrings_test.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+def dataset_with_empty_string():
+    return TextData(
+        raw_text=[
+            ''
+        ]
+    )
+
+# =================
+# ----- Tests -----
+# =================
```
Comment (Nadav-Barak) [inline tests/nlp/checks/data_integrity/frequent_substrings_test.py:74]: Additional comment - add 1-2 tests about the return value of the check (currently all the tests check for the result of the condition)

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+        return ngram_info, qualifying_samples
+
+    def _get_frequent_ngrams(self, data):
+        n = self.min_ngram
+        final_results = {}
+        num_samples = len(data)
+        while data:
+            result, qualifying_samples = self._get_ngram_frequencies_and_indexes(data, n, num_samples)
+            result = {ngram: info for ngram, info in result.items() if info['freq'] >= self.min_threshold}
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:117]: why filter here for ` >= self.min_threshold`? Couldn't we do that inside the for ngram in ngram_info: loop?

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+        self.epsilon = epsilon
+        self.significant_threshold = significant_threshold
+
+    @staticmethod
+    def _extract_ngrams(text, n):
+        words = text.split()
+        return [' '.join(words[i:i + n]) for i in range(len(words) - n + 1)]
+
+    @staticmethod
+    def _get_peak_cut_ind(df, min_diff=0.05):
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:77]: min diff should also be a check param

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        df['diff'] = df['Frequency'].diff()
+        df[df['diff'] <= min_diff]['diff'] = 0
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:79]: I'm considering whether the diff should be on on the frequency (like you did here) or rather on the change in frequency  - meaning `(freq_i - freq_i_+1) / freq_i`. What do you think?

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+
+            df = self._filter_significant(df)
+            df['Number of Samples'] = df['Sample IDs'].str.len()
+
+            value = df
+            percent_of_frequent = sum(df['Number of Samples'])
+            display = [
+                f'{format_percent(percent_of_frequent)} of data samples share common substrings.',
+                'Each row in the table shows an example of a frequent substring and the number of times it appears.',
+                df[['Text', 'Number of Samples', 'Frequency']].iloc[slice(0, self.n_to_show)]
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:202]: For display, better display frequency in percents and change its name accordingly.

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+            ]
+
+        return CheckResult(
+            value=value,
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:206]: Maybe better format the value as a dict? It can have a very long list of IDs for 'Sample IDs'

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+        """
+        def condition(result: Dict) -> ConditionResult:
+            num_substrings = len(result)
+            msg = f'Found {num_substrings} substrings with ratio above threshold'
+            if num_substrings >= min_substrings:
+                return ConditionResult(ConditionCategory.WARN, msg)
+            else:
+                return ConditionResult(ConditionCategory.PASS, msg)
+
+        return self.add_condition(f'There should be not more than {min_substrings} frequent substrings',
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:226]:
```suggestion
        return self.add_condition(f'No more than {min_substrings} frequent substrings',
```
Better to shorten it for display in suites.

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+                'Each row in the table shows an example of a frequent substring and the number of times it appears.',
+                df[['Text', 'Number of Samples', 'Frequency']].iloc[slice(0, self.n_to_show)]
+            ]
+
+        return CheckResult(
+            value=value,
+            display=display
+        )
+
+    def add_condition_zero_result(self, min_substrings: int = 1):
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:210]: Modify to also include the frequency parameter, not more the X substrings with frequency over Y

Can see something similar at deepchecks/utils/abstracts/feature_drift.py

**Review by noamzbr — COMMENTED** [review]
Great work! 🚀 
Apart from the comments themselves - one meta comment is that it would be very helpful to add much more comments and dosctrings to your private methods, as they are pretty complex and sometimes hard to follow.

Another thing that needs doing before merging - adding imports for the new checks in the different __init__ files of the checks folder. Look at the __init__ files inside data_integrity and in the top level of the check folder.
Finally, we should also add the check to the default data integrity suite, with it's default condition.

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+                    ngram_info[ngram]['original_indexes'].add(item[0])
+                    ngram_info[ngram]['new_indexes'].add(index)
+
+        for ngram in ngram_info:
+            ngram_info[ngram]['original_indexes'] = list(ngram_info[ngram]['original_indexes'])
+            ngram_info[ngram]['new_indexes'] = list(ngram_info[ngram]['new_indexes'])
+            ngram_info[ngram]['freq'] /= num_samples
+            if ngram_info[ngram]['freq'] >= self.min_threshold:
+                for index in ngram_info[ngram]['new_indexes']:
+                    qualifying_samples.add(index)
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/checks/data_integrity/frequent_substrings.py]: used to narrow down the samples that contain the ngrams of lengh n, so that we can look only at them for ngrams of length n+1

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 24]
```diff
@@ -0,0 +1,417 @@
...
+            Input dataframe with a 'Frequency' column.
+
+        Returns
+        -------
+        int
+            The index of the maximum relative change in the 'Frequency' column. If no significant peak is
+            found, returns the length of the dataframe.
+        """
+        diff = df['Frequency'].diff().abs()
+        ratio = diff.div(df['Frequency'].shift(1))
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:290]: sure .shift(1) is the current value and not the next value? 
To rephrase, asking if it's (t[i]-t[i-1]) / t[i] or (t[i]-t[i-1]) / t[i-1]. I think that it should be the first one.

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 25]
```diff
@@ -0,0 +1,417 @@
...
+            df = pd.DataFrame({
+                'Text': [item[0] for item in sorted_substrings],
+                'Frequency': [item[1]['freq'] for item in sorted_substrings],
+                'Sample IDs': [item[1]['original_indexes'] for item in sorted_substrings]
+            })
+
+            df = self._filter_peak_and_significant(df)
+
+            if self.n_samples < num_samples:
+                dataset = context.get_data_by_kind(dataset_kind)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:367]: if we didn't override the dataset with the sampled one we could have avoided this

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 26]
```diff
@@ -0,0 +1,417 @@
...
+__all__ = ['FrequentSubstrings']
+
+
+class FrequentSubstrings(SingleDatasetCheck):
+    """Checks for frequent substrings in the dataset.
+
+    From the given dataset, substrings of varying lengths (n-grams) are extracted.
+    The frequencies of these n-grams are calculated and only substrings exceeding a defined minimum length are retained.
+    The substrings are then sorted by their frequencies and the most frequent substrings are identified.
+    Finally, peak substrings and those surpassing a significance level are displayed.
```
[the remark this answers is not in the document]
Comment (shiritdvir) [reply deepchecks/nlp/checks/data_integrity/frequent_substrings.py:33]: The substring at the peak (before the drop)

[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/checks/data_integrity/frequent_substrings.py:33]: I meant that this won't be clear to the user (:

File: deepchecks/nlp/checks/data_integrity/frequent_substrings.py [hunk 27]
```diff
@@ -0,0 +1,472 @@
...
+    min_relative_change : float, optional, default=0.05
+        Defines the threshold for relative change. If the computed relative
+        change falls below this specified threshold, it is
+        considered insignificant and is thus set to zero.
+
+    """
+
+    def __init__(
+        self,
+        n_to_show: int = 5,
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/frequent_substrings.py:65]: for consistency, have all of these with no spaces.

**Review by noamzbr — APPROVED** [review]
